File: internal/capability/ssh_test.go
package capability

import (
	"context"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/ro-ag/ptrack/internal/model"
)

type recordingProcessRunner struct {
	spec    ProcessSpec
	result  ProcessResult
	err     error
	inspect func(ProcessSpec)
}

func (r *recordingProcessRunner) Run(_ context.Context, spec ProcessSpec) (ProcessResult, error) {
	r.spec = spec
	if r.inspect != nil {
		r.inspect(spec)
	}
	return r.result, r.err
}

func TestSSHCommandUsesPinnedKeyAgentOnlyAndExactCommand(t *testing.T) {
	key := "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA=="
	capability, now := approvedSSH(t, &model.SSHScope{
		Host: "example.com", Port: 2222, User: "deploy", HostKey: key,
		RemoteCommands: []string{"printf safe"},
	})
	runner := &recordingProcessRunner{result: ProcessResult{Stdout: "safe"}}
	executor := SSHExecutor{Runner: runner, Now: func() time.Time { return now }}
	result, err := executor.Execute(context.Background(), capability, "agent-codex", t.TempDir(), SSHRequest{
		Operation: SSHRemoteCommand, Command: "printf safe",
	})
	if err != nil || result.Stdout != "safe" {
		t.Fatalf("result=%+v err=%v", result, err)
	}
	joined := strings.Join(runner.spec.Args, " ")
	for _, required := range []string{"BatchMode=yes", "PasswordAuthentication=no", "StrictHostKeyChecking=yes", "UserKnownHostsFile=", "-p 2222", "deploy@example.com", "printf safe"} {
		if !strings.Contains(joined, required) {
			t.Errorf("argv missing %q: %v", required, runner.spec.Args)
		}
	}
	if strings.Contains(joined, "IdentityFile") || runner.spec.Name != "ssh" {
		t.Errorf("unexpected process: %+v", runner.spec)
	}
}

func TestSSHUploadConfinesLocalRemotePathsAndUsesDirectSCPArgs(t *testing.T) {
	project := t.TempDir()
	if err := os.Mkdir(filepath.Join(project, "dist"), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(project, "dist", "app.js"), []byte("x"), 0o600); err != nil {
		t.Fatal(err)
	}
	key := "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA=="
	capability, now := approvedSSH(t, &model.SSHScope{
		Host: "example.com", User: "deploy", HostKey: key,
		AllowUpload: true, UploadRoots: []string{"dist"}, UploadRemoteRoots: []string{"/srv/app"},
	})
	var stagedSource string
	runner := &recordingProcessRunner{inspect: func(spec ProcessSpec) {
		separator := indexOf(spec.Args, "--")
		if separator < 0 || separator+1 >= len(spec.Args) {
			t.Fatalf("scp args = %v", spec.Args)
		}
		stagedSource = spec.Args[separator+1]
		contents, readErr := os.ReadFile(stagedSource)
		if readErr != nil || string(contents) != "x" {
			t.Fatalf("staged upload = %q, %v", contents, readErr)
		}
		if stagedSource == filepath.Join(project, "dist", "app.js") {
			t.Fatal("scp received the mutable project pathname")
		}
	}}
	executor := SSHExecutor{Runner: runner, Now: func() time.Time { return now }}
	_, err := executor.Execute(context.Background(), capability, "agent-codex", project, SSHRequest{
		Operation: SSHUpload, LocalPath: "dist/app.js", RemotePath: "/srv/app/app.js",
	})
	if err != nil {
		t.Fatal(err)
	}
	if runner.spec.Name != "scp" || !contains(runner.spec.Args, "--") || !contains(runner.spec.Args, "deploy@example.com:/srv/app/app.js") {
		t.Fatalf("scp spec = %+v", runner.spec)
	}
	if _, err := os.Stat(stagedSource); !errors.Is(err, os.ErrNotExist) {
		t.Fatalf("private upload staging file remains: %v", err)
	}
	if _, err := executor.Execute(context.Background(), capability, "agent-codex", project, SSHRequest{
		Operation: SSHUpload, LocalPath: "dist/app.js", RemotePath: "/etc/passwd",
	}); err == nil {
		t.Fatal("remote upload escape authorized")
	}
}

func TestSSHTransfersEnforceByteLimitsAndInstallDownloadsSafely(t *testing.T) {
	project := t.TempDir()
	if err := os.MkdirAll(filepath.Join(project, "dist"), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(filepath.Join(project, "dist", "large.bin"), []byte("too-large"), 0o600); err != nil {
		t.Fatal(err)
	}
	key := "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA=="
	upload, now := approvedCapability(t, model.Capability{
		Name: "upload", Kind: model.CapabilitySSH, AgentProfile: "agent-codex",
		Limits: model.CapabilityLimits{MaxRequestBytes: 4},
		SSH: &model.SSHScope{
			Host: "example.com", User: "deploy", HostKey: key,
			AllowUpload: true, UploadRoots: []string{"dist"}, UploadRemoteRoots: []string{"/srv/app"},
		},
	})
	executor := SSHExecutor{Runner: &recordingProcessRunner{}, Now: func() time.Time { return now }}
	if _, err := executor.Execute(context.Background(), upload, "agent-codex", project, SSHRequest{
		Operation: SSHUpload, LocalPath: "dist/large.bin", RemotePath: "/srv/app/large.bin",
	}); ClassifySSHError(err, "") != "request-limit" {
		t.Fatalf("oversized upload error = %v", err)
	}

	if err := os.MkdirAll(filepath.Join(project, "artifacts"), 0o755); err != nil {
		t.Fatal(err)
	}
	download, _ := approvedCapability(t, model.Capability{
		Name: "download", Kind: model.CapabilitySSH, AgentProfile: "agent-codex",
		Limits: model.CapabilityLimits{MaxResponseBytes: 16},
		SSH: &model.SSHScope{
			Host: "example.com", User: "deploy", HostKey: key,
			AllowDownload: true, DownloadRoots: []string{"artifacts"}, DownloadRemoteRoots: []string{"/srv/releases"},
		},
	})
	runner := &recordingProcessRunner{result: ProcessResult{Stdout: "release"}}
	executor.Runner = runner
	if _, err := executor.Execute(context.Background(), download, "agent-codex", project, SSHRequest{
		Operation: SSHDownload, LocalPath: "artifacts/release.txt", RemotePath: "/srv/releases/release.txt",
	}); err != nil {
		t.Fatal(err)
	}
	contents, err := os.ReadFile(filepath.Join(project, "artifacts", "release.txt"))
	if err != nil || string(contents) != "release" {
		t.Fatalf("installed download = %q, %v", contents, err)
	}
	if runner.spec.Name != "ssh" || runner.spec.MaxOutputBytes != 16 ||
		!contains(runner.spec.Args, "cat -- /srv/releases/release.txt") {
		t.Fatalf("bounded download spec = %+v", runner.spec)
	}
	runner.result = ProcessResult{Stdout: strings.Repeat("x", 16), Truncated: true}
	if result, err := executor.Execute(context.Background(), download, "agent-codex", project, SSHRequest{
		Operation: SSHDownload, LocalPath: "artifacts/oversized.txt", RemotePath: "/srv/releases/oversized.txt",
	}); err == nil || result.Stdout != "" {
		t.Fatalf("unbounded download result=%+v err=%v", result, err)
	}
}

func TestSSHDownloadRejectsDestinationSymlinkSwapAfterTransfer(t *testing.T) {
	project := t.TempDir()
	outside := t.TempDir()
	artifacts := filepath.Join(project, "artifacts")
	if err := os.Mkdir(artifacts, 0o755); err != nil {
		t.Fatal(err)
	}
	key := "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA=="
	capability, now := approvedSSH(t, &model.SSHScope{
		Host: "example.com", User: "deploy", HostKey: key,
		AllowDownload: true, DownloadRoots: []string{"artifacts"}, DownloadRemoteRoots: []string{"/srv/releases"},
	})
	runner := &recordingProcessRunner{result: ProcessResult{Stdout: "secret"}, inspect: func(spec ProcessSpec) {
		if err := os.Rename(artifacts, filepath.Join(project, "artifacts-original")); err != nil {
			t.Fatal(err)
		}
		if err := os.Symlink(outside, artifacts); err != nil {
			t.Fatal(err)
		}
	}}
	executor := SSHExecutor{Runner: runner, Now: func() time.Time { return now }}
	_, err := executor.Execute(context.Background(), capability, "agent-codex", project, SSHRequest{
		Operation: SSHDownload, LocalPath: "artifacts/release.txt", RemotePath: "/srv/releases/release.txt",
	})
	if ClassifySSHError(err, "") != "denied" {
		t.Fatalf("symlink swap error = %v", err)
	}
	if _, statErr := os.Stat(filepath.Join(outside, "release.txt")); !errors.Is(statErr, os.ErrNotExist) {
		t.Fatalf("download escaped project: %v", statErr)
	}
}

func indexOf(values []string, wanted string) int {
	for index, value := range values {
		if value == wanted {
			return index
		}
	}
	return -1
}

func TestSSHForwardingIsLoopbackAndDirectionScoped(t *testing.T) {
	key := "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA=="
	capability, now := approvedSSH(t, &model.SSHScope{
		Host: "example.com", User: "deploy", HostKey: key,
		LocalForwardTargets: []string{"db.internal:5432"},
	})
	runner := &recordingProcessRunner{}
	executor := SSHExecutor{Runner: runner, Now: func() time.Time { return now }}
	_, err := executor.Execute(context.Background(), capability, "agent-codex", t.TempDir(), SSHRequest{
		Operation: SSHLocalForward, ForwardTarget: "db.internal:5432", ListenPort: 15432,
	})
	if err != nil {
		t.Fatal(err)
	}
	joined := strings.Join(runner.spec.Args, " ")
	if !strings.Contains(joined, "-L 127.0.0.1:15432:db.internal:5432") {
		t.Fatalf("forward argv = %v", runner.spec.Args)
	}
	if _, err := executor.Execute(context.Background(), capability, "agent-codex", t.TempDir(), SSHRequest{
		Operation: SSHRemoteForward, ForwardTarget: "db.internal:5432", ListenPort: 15432,
	}); err == nil {
		t.Fatal("local forward approval implied remote forwarding")
	}
}

func TestClassifySSHErrorUsesSanitizedClasses(t *testing.T) {
	for diagnostic, want := range map[string]string{
		"ssh: Could not resolve hostname x":                "dns",
		"Host key verification failed":                     "host-key",
		"Permission denied (publickey)":                    "authentication",
		"channel open failed: administratively prohibited": "remote-policy",
	} {
		if got := ClassifySSHError(errors.New("exit status 255"), diagnostic); got != want {
			t.Errorf("%q => %q want %q", diagnostic, got, want)
		}
	}
}

File: crates/ptrack-capability/src/ssh_test.rs
use std::ffi::OsString;
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;

use ptrack_core::{CapabilityKind, SshScope};
use tokio_util::sync::CancellationToken;

use super::AuditRecorder;
use super::git::{ProcessRunner, RunFuture};
use super::process::{ProcessResult, ProcessSpec};
#[cfg(unix)]
use super::ssh::unlink_owned_download_temp;
use super::ssh::{SshExecutor, SshRequest, classify_ssh_exit};
use super::test_support::{TempDir, draft, refresh_approval};

pub(super) fn assert_cap_077_through_084_ssh_contract() {
    assert_eq!(
        classify_ssh_exit("Host key verification failed"),
        "host-key"
    );
    assert_eq!(
        classify_ssh_exit("Permission denied (publickey)"),
        "authentication"
    );
    assert_eq!(classify_ssh_exit("No route to host"), "routing");
    assert_eq!(
        classify_ssh_exit("Administratively prohibited"),
        "remote-policy"
    );
}

#[tokio::test]
async fn ssh_exact_argv_and_noncomposable_grants_deny_before_spawn() {
    let temp = TempDir::new("ssh-command");
    let capability = ssh_capability(|scope| {
        scope.remote_commands = vec!["printf safe".to_owned()];
    });
    let runner = RecordingRunner::success(Vec::new());
    let executor = SshExecutor::from_parts(AuditRecorder::new(None), &runner);
    let denied = executor
        .execute(
            &CancellationToken::new(),
            &capability,
            "agent-codex",
            temp.path(),
            &SshRequest {
                operation: "remote-command".to_owned(),
                command: "printf safe; touch escaped".to_owned(),
                local_path: String::new(),
                remote_path: String::new(),
                forward_target: String::new(),
                listen_port: 0,
            },
        )
        .await
        .unwrap_err();
    assert_eq!(denied.class(), "denied");
    assert!(runner.calls.lock().unwrap().is_empty());

    executor
        .execute(
            &CancellationToken::new(),
            &capability,
            "agent-codex",
            temp.path(),
            &SshRequest {
                operation: "remote-command".to_owned(),
                command: "printf safe".to_owned(),
                local_path: String::new(),
                remote_path: String::new(),
                forward_target: String::new(),
                listen_port: 0,
            },
        )
        .await
        .unwrap();
    let calls = runner.calls.lock().unwrap();
    let args = strings(&calls[0].args);
    assert_eq!(calls[0].name, OsString::from("ssh"));
    let null_config = if cfg!(windows) { "NUL" } else { "/dev/null" };
    assert_eq!(&args[..4], ["-F", null_config, "-o", "BatchMode=yes"]);
    assert!(
        args.windows(2)
            .any(|pair| pair == ["-o", "PasswordAuthentication=no"])
    );
    assert!(
        args.windows(2)
            .any(|pair| pair == ["-o", "StrictHostKeyChecking=yes"])
    );
    assert_eq!(
        &args[args.len() - 3..],
        ["-T", "deploy@example.test", "printf safe"]
    );
}

#[tokio::test]
async fn upload_is_immutable_bounded_staged_copy_and_scp_is_direct() {
    let temp = TempDir::new("ssh-upload");
    fs::create_dir(temp.path().join("uploads")).unwrap();
    fs::write(temp.path().join("uploads/source.txt"), b"payload").unwrap();
    let capability = ssh_capability(|scope| {
        scope.allow_upload = true;
        scope.upload_roots = vec!["uploads".to_owned()];
        scope.upload_remote_roots = vec!["/srv/uploads".to_owned()];
    });
    let runner = RecordingRunner::success(Vec::new());
    let result = SshExecutor::from_parts(AuditRecorder::new(None), &runner)
        .execute(
            &CancellationToken::new(),
            &capability,
            "agent-codex",
            temp.path(),
            &SshRequest {
                operation: "upload".to_owned(),
                command: String::new(),
                local_path: "uploads/source.txt".to_owned(),
                remote_path: "/srv/uploads/file.txt".to_owned(),
                forward_target: String::new(),
                listen_port: 0,
            },
        )
        .await
        .unwrap();
    assert_eq!(result.diagnostic, "none");
    let calls = runner.calls.lock().unwrap();
    assert_eq!(calls[0].name, OsString::from("scp"));
    let args = strings(&calls[0].args);
    assert!(args.contains(&"-P".to_owned()));
    assert_eq!(args[args.len() - 3], "--");
    assert_eq!(
        args.last().unwrap(),
        "deploy@example.test:/srv/uploads/file.txt"
    );
    assert_eq!(
        runner.staged.lock().unwrap().as_deref(),
        Some(b"payload".as_slice())
    );
}

#[tokio::test]
async fn download_payload_is_not_returned_and_parent_swap_fails_closed() {
    let temp = TempDir::new("ssh-download");
    fs::create_dir(temp.path().join("downloads")).unwrap();
    let capability = ssh_capability(|scope| {
        scope.allow_download = true;
        scope.download_roots = vec!["downloads".to_owned()];
        scope.download_remote_roots = vec!["/srv/downloads".to_owned()];
    });
    let runner = RecordingRunner::success(b"downloaded".to_vec());
    let result = SshExecutor::from_parts(AuditRecorder::new(None), &runner)
        .execute(
            &CancellationToken::new(),
            &capability,
            "agent-codex",
            temp.path(),
            &download_request(),
        )
        .await
        .unwrap();
    assert!(result.stdout.is_empty());
    assert_eq!(
        fs::read(temp.path().join("downloads/result.bin")).unwrap(),
        b"downloaded"
    );

    #[cfg(unix)]
    {
        let outside = temp.path().join("outside");
        fs::create_dir(&outside).unwrap();
        let runner =
            RecordingRunner::with_swap(b"hostile".to_vec(), temp.path().join("downloads"), outside);
        let error = SshExecutor::from_parts(AuditRecorder::new(None), &runner)
            .execute(
                &CancellationToken::new(),
                &capability,
                "agent-codex",
                temp.path(),
                &download_request(),
            )
            .await
            .unwrap_err();
        assert_eq!(error.class(), "denied");
    }
}

#[tokio::test]
async fn transfer_remote_paths_are_exact_normalized_shell_safe_and_deny_before_spawn() {
    let temp = TempDir::new("ssh-remote-paths");
    fs::create_dir(temp.path().join("uploads")).unwrap();
    fs::create_dir(temp.path().join("downloads")).unwrap();
    fs::write(temp.path().join("uploads/source.txt"), b"payload").unwrap();
    let capability = ssh_capability(|scope| {
        scope.allow_upload = true;
        scope.allow_download = true;
        scope.upload_roots = vec!["uploads".to_owned()];
        scope.download_roots = vec!["downloads".to_owned()];
        scope.upload_remote_roots = vec!["/srv/uploads".to_owned()];
        scope.download_remote_roots = vec!["/srv/downloads".to_owned()];
    });
    let hostile = [
        "/srv/uploads/file;touch-pwned",
        "/srv/uploads/two words",
        "/srv/uploads/$(touch-pwned)",
        "/srv/uploads/`touch-pwned`",
        "/srv/uploads/'quoted'",
        "/srv/uploads/\"quoted\"",
        "/srv/uploads/control\nname",
        "/srv/uploads/../outside",
        "/srv/uploads//file",
        "/srv/uploads/./file",
    ];
    for &remote_path in &hostile {
        let runner = RecordingRunner::success(Vec::new());
        let error = SshExecutor::from_parts(AuditRecorder::new(None), &runner)
            .execute(
                &CancellationToken::new(),
                &capability,
                "agent-codex",
                temp.path(),
                &SshRequest {
                    operation: "upload".to_owned(),
                    command: String::new(),
                    local_path: "uploads/source.txt".to_owned(),
                    remote_path: remote_path.to_owned(),
                    forward_target: String::new(),
                    listen_port: 0,
                },
            )
            .await
            .unwrap_err();
        assert_eq!(error.class(), "denied");
        assert!(runner.calls.lock().unwrap().is_empty(), "{remote_path:?}");
    }
    for remote_path in hostile.map(|path| path.replacen("uploads", "downloads", 1)) {
        let runner = RecordingRunner::success(Vec::new());
        let error = SshExecutor::from_parts(AuditRecorder::new(None), &runner)
            .execute(
                &CancellationToken::new(),
                &capability,
                "agent-codex",
                temp.path(),
                &SshRequest {
                    operation: "download".to_owned(),
                    command: String::new(),
                    local_path: "downloads/result.bin".to_owned(),
                    remote_path: remote_path.clone(),
                    forward_target: String::new(),
                    listen_port: 0,
                },
            )
            .await
            .unwrap_err();
        assert_eq!(error.class(), "denied");
        assert!(runner.calls.lock().unwrap().is_empty(), "{remote_path:?}");
    }
}

#[cfg(unix)]
#[tokio::test]
async fn download_install_cleans_owned_temps_without_deleting_replacements() {
    use std::fs::OpenOptions;
    use std::os::fd::AsFd as _;

    let temp = TempDir::new("ssh-download-cleanup");
    let downloads = temp.path().join("downloads");
    fs::create_dir(&downloads).unwrap();
    let mut capability = ssh_capability(|scope| {
        scope.allow_download = true;
        scope.download_roots = vec!["downloads".to_owned()];
        scope.download_remote_roots = vec!["/srv/downloads".to_owned()];
    });
    capability.limits.max_response_bytes = 4;
    capability = refresh_approval(capability);
    let error = SshExecutor::from_parts(
        AuditRecorder::new(None),
        &RecordingRunner::success(b"too-large".to_vec()),
    )
    .execute(
        &CancellationToken::new(),
        &capability,
        "agent-codex",
        temp.path(),
        &download_request(),
    )
    .await
    .unwrap_err();
    assert_eq!(error.class(), "response-limit");
    assert_no_download_temps(&downloads);

    fs::create_dir(downloads.join("result.bin")).unwrap();
    capability.limits.max_response_bytes = 1024;
    capability = refresh_approval(capability);
    let error = SshExecutor::from_parts(
        AuditRecorder::new(None),
        &RecordingRunner::success(b"payload".to_vec()),
    )
    .execute(
        &CancellationToken::new(),
        &capability,
        "agent-codex",
        temp.path(),
        &download_request(),
    )
    .await
    .unwrap_err();
    assert_eq!(error.class(), "internal");
    assert_no_download_temps(&downloads);

    let directory = OpenOptions::new().read(true).open(&downloads).unwrap();
    let owned_path = downloads.join(".ptrack-download-owned");
    fs::write(&owned_path, b"owned").unwrap();
    let owned = OpenOptions::new().read(true).open(&owned_path).unwrap();
    fs::rename(&owned_path, downloads.join("moved-owned")).unwrap();
    fs::write(&owned_path, b"canary").unwrap();
    unlink_owned_download_temp(directory.as_fd(), ".ptrack-download-owned", owned.as_fd());
    assert_eq!(fs::read(&owned_path).unwrap(), b"canary");
}

#[cfg(unix)]
fn assert_no_download_temps(directory: &std::path::Path) {
    assert!(fs::read_dir(directory).unwrap().all(|entry| {
        !entry
            .unwrap()
            .file_name()
            .to_string_lossy()
            .starts_with(".ptrack-download-")
    }));
}

fn ssh_capability(edit: impl FnOnce(&mut SshScope)) -> ptrack_core::Capability {
    let mut capability = draft(CapabilityKind::Ssh);
    let mut scope = SshScope {
        alias: "deploy".to_owned(),
        host: "example.test".to_owned(),
        port: 22,
        user: "deploy".to_owned(),
        host_key: "ssh-ed25519 QUJDREVGR0hJSktMTU5PUA==".to_owned(),
        allow_git: false,
        remote_commands: Vec::new(),
        allow_upload: false,
        allow_download: false,
        upload_roots: Vec::new(),
        download_roots: Vec::new(),
        upload_remote_roots: Vec::new(),
        download_remote_roots: Vec::new(),
        allow_interactive_shell: false,
        local_forward_targets: Vec::new(),
        remote_forward_targets: Vec::new(),
    };
    edit(&mut scope);
    capability.ssh = Some(scope);
    refresh_approval(capability)
}

fn download_request() -> SshRequest {
    SshRequest {
        operation: "download".to_owned(),
        command: String::new(),
        local_path: "downloads/result.bin".to_owned(),
        remote_path: "/srv/downloads/result.bin".to_owned(),
        forward_target: String::new(),
        listen_port: 0,
    }
}

#[derive(Clone, Debug)]
struct CapturedSpec {
    name: OsString,
    args: Vec<OsString>,
}

struct RecordingRunner {
    calls: Mutex<Vec<CapturedSpec>>,
    stdout: Vec<u8>,
    staged: Mutex<Option<Vec<u8>>>,
    #[cfg(unix)]
    swap: Option<(PathBuf, PathBuf)>,
}

impl RecordingRunner {
    fn success(stdout: Vec<u8>) -> Self {
        Self {
            calls: Mutex::new(Vec::new()),
            stdout,
            staged: Mutex::new(None),
            #[cfg(unix)]
            swap: None,
        }
    }

    #[cfg(unix)]
    fn with_swap(stdout: Vec<u8>, path: PathBuf, target: PathBuf) -> Self {
        Self {
            calls: Mutex::new(Vec::new()),
            stdout,
            staged: Mutex::new(None),
            swap: Some((path, target)),
        }
    }
}

impl ProcessRunner for RecordingRunner {
    fn run<'a>(
        &'a self,
        spec: &'a ProcessSpec,
        _cancellation: &'a CancellationToken,
    ) -> RunFuture<'a> {
        Box::pin(async move {
            self.calls.lock().unwrap().push(CapturedSpec {
                name: spec.name.clone(),
                args: spec.args.clone(),
            });
            if spec.name == "scp" {
                let path = PathBuf::from(&spec.args[spec.args.len() - 2]);
                *self.staged.lock().unwrap() = fs::read(path).ok();
            }
            #[cfg(unix)]
            if let Some((path, target)) = &self.swap {
                let old = path.with_extension("old");
                fs::rename(path, &old).unwrap();
                std::os::unix::fs::symlink(target, path).unwrap();
            }
            Ok(ProcessResult {
                exit_code: 0,
                stdout: self.stdout.clone(),
                stderr: Vec::new(),
                truncated: false,
            })
        })
    }
}

fn strings(values: &[OsString]) -> Vec<String> {
    values
        .iter()
        .map(|value| value.to_string_lossy().into_owned())
        .collect()
}
